pkg/controller/container-runtime-config/container_runtime_config_controller.go

```diff
@@ -207,9 +225,13 @@ func (ctrl *Controller) Run(workers int, stopCh <-chan struct{}) {
 	defer utilruntime.HandleCrash()
 	defer ctrl.queue.ShutDown()
 	defer ctrl.imgQueue.ShutDown()
-
-	if !cache.WaitForCacheSync(stopCh, ctrl.mcpListerSynced, ctrl.mccrListerSynced, ctrl.ccListerSynced,
-		ctrl.imgListerSynced, ctrl.icspListerSynced, ctrl.idmsListerSynced, ctrl.itmsListerSynced, ctrl.clusterVersionListerSynced) {
+	listerCaches := []cache.InformerSynced{ctrl.mcpListerSynced, ctrl.mccrListerSynced, ctrl.ccListerSynced,
+		ctrl.imgListerSynced, ctrl.icspListerSynced, ctrl.idmsListerSynced, ctrl.itmsListerSynced, ctrl.clusterVersionListerSynced}
+	if ctrl.sigstoreAPIEnabled() {
+		ctrl.enableClusterImagePolicyHandler()
+		listerCaches = append(listerCaches, ctrl.clusterImagePolicyListerSynced)
```

Comment:
https://github.com/openshift/machine-config-operator/pull/4160#discussion_r1495230009 reports that the feature gate is flipped on only in the sync loop.

At a first glance, the value flipping seems to contradict https://github.com/openshift/machine-config-operator/pull/4160#discussion_r1487541546 , but if I understand the code correctly, `cmd/machine-config-controller.createControllers` calls this file’s `New` _before_ `FeatureGateAccess.InitialFeatureGatesObserved`.

So there is no mystery … but, in practice, that also means that this wait for cache sync on `clusterImagePolicyListerSynced` is very unlikely to trigger. Is that an issue?

Reply:
My mistake, this is `Run`, not `New`…

… but in that case we should already have seen `InitialFeatureGatesObserved`, so the gate should be on before we first run `syncImageConfig`.

@QiWang19 was the `panic` in https://github.com/openshift/machine-config-operator/pull/4160#discussion_r1495230009 triggered when removing also the `enableClusterImagePolicyHandler`? Or is the initial wait for `InitialFeatureGatesObserved` timing out, causing us to only observe the feature gate value later? Something else?

---

If the _first_ time we are observing the feature gate to on really is inside the sync loop, it seems to me that https://github.com/openshift/machine-config-operator/pull/4160#discussion_r1480627820 applies again, and the availability of the sync gate value should trigger a resync and `syncImageConfig` to be called soon.

OTOH the design of the feature gate, with the `ForceExit` default behavior, suggests that observing the feature gate is not a typical pattern, so I’m quite likely missing something.

Reply:
I am thinking that something in the bootstrap e2e [test](https://github.com/openshift/machine-config-operator/blob/39334cf103f699e196fcada236b65d5993f469bc/test/e2e-bootstrap/bootstrap_test.go#L419) is missing.  I did not see it called `InitialFeatureGatesObserved`.  Rechecked the bootstrap e2e test [log](https://storage.googleapis.com/test-platform-results/pr-logs/pull/openshift_machine-config-operator/4160/pull-ci-openshift-machine-config-operator-master-bootstrap-unit/1758227788193599488/build-log.txt), 
before the `syncImageConfig`, the featuregate was not ready, the code assumed it as disabled in the `Run()` and `New()`. So the first time we are observing the feature gate enabled is inside the sync loop.
I understand the actual implementation is that when the ContainerRuntimeConfigController gets `Run()`, the feature gate should be ready with waiting for `InitialFeatureGatesObserved` and that observing the feature gate is not a typical pattern.

Reply:
on second thought, maybe we can’t simplify that much?

If the one-minute wait for feature gate status times out, we still don’t have a value, and the feature gate might hypothetically be available and set to true later. Do we need to observe that event and trigger a resync?

Or maybe, should we block for the feature gate indefinitely, so that we always have a value? The `WaitForCacheSync` loop does seem to block indefinitely…

(Warning: I don’t understand any of this well.)

Reply:
@mtrmac  I think we can simplify that much. When running `*Controller.Run()` the featuregates are always have a value. 
if the featuregate not observed during one-minute wait before `Run()`, it will exit.  
https://github.com/openshift/machine-config-operator/blob/398c1710f12622e9c0f94f99b12abeabed3ac5f7/cmd/machine-config-controller/start.go#L106-L113

Reply:
Oh, I have completely missed that. Thanks for pointing that out!

So, after all, we always have a feature gate value, so we don’t have to worry about the feature gate state being unknown (and generating incorrect `policy.json`), nor about the feature gate state changing at runtime.

Reply:
> I am thinking that something in the bootstrap e2e [test](https://github.com/openshift/machine-config-operator/blob/39334cf103f699e196fcada236b65d5993f469bc/test/e2e-bootstrap/bootstrap_test.go#L419)

regarding the abvoe thought,
I tested bootstrap-e2e with following changes: 
- add one minute wait for featuregate in the `Run()` , 
- and set `ctrl.clusterImagePolicyLister` when this featuregate enabled.
- ask the ci to `/test bootstrap-unit`

The test passed and the log showed featuregate observed in the `Run()`, no panic when it goes to the line `ctrl.clusterImagePolicyLister.List(labels.Everything())` in the `syncImageConfig()`. so somewhere in the test setup needs fix to follow the logic in the actual code.